### ISSUE TITLE
An explanation of DAGMC plot artifacts

### ISSUE THREAD

**Issue by pshriwise** [openmc-dev/openmc#2660]
**TL;DR:** Artifacts sometimes appear in the plotter in regions with unmerged surfaces in DAGMC models. The order in which we check volumes/cells during the `find_cell` operation matters in these cases and we should defer checking DAGMC's implicit complement until the end.

---
## The problem

"Streaks" sometimes appear in plotting images of DAGMC models. As in our CSG geometry, this can occur if the slicing plane is on the boundary of a region, but these artifacts also appear throughout volumes in DAGMC universes. This is an explanation as to why this occurs and a proposal for a fix.

![image](https://github.com/openmc-dev/openmc/assets/4563941/2cdc817b-e611-46e8-bb1e-2165ecc67533)

Image from the Copenhagen Atomics open [MSRE model](https://github.com/openmsr/msre)

I can confirm that there are unmerged, coincident surfaces in this model in this region that are causing these problems. It's possible that other models have different causes for these artifacts, but here's my explanation of why it occurs for unmerged, coincident surfaces.

## DAGMC model description

Let's say a model is created with the following two volumes, 1 and 2.

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/bed4820e-d7d3-48d4-b07a-7412e083400e">

Ideally, there would be a single surface, surface 1, between the two volumes. It would have a "forward" sense with respect to volume 1 and a "reverse" sense with respect to volume 2. This sense is used to determine whether a particle crossing is entering or exiting one of the volumes.

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/0e540d7b-6562-4627-80c3-a5368d82ff23">

This shared surface is generated as part of the "imprint and merge" process in the CAD engine. Part of this process is to combine, or merge, coincident surfaces into one -- resulting in surface 1 in this case. In the case that this process fails, the model has two coincident surfaces, surfaces 1 and 2, between volumes 1 and 2. In this case, we get the following:

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/c4bcd327-0e5e-407f-af21-3948b57b8134">

The surfaces are shown with some separation to ease visualization, but they are numerically coincident with one another. Surface 1 now has a "forward" sense with respect to volume 1 and nothing on the other side. Surface 2 has a "forward" sense with respect to volume 2 as well and nothing on the other side. When building the [implicit complement](https://svalinn.github.io/DAGMC/contribute/devtheory.html#implicit-complement) in DAGMC, the programatically constructed space claims the "nothing" side of these surfaces. A key thing to remember is that _both_ of these surfaces are now part of the IPC volume generated by 

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/1c07b283-13e5-4bb5-b332-1a72f90704f1">

At first glance, it might appear that the implicit complement occupies the space between the two unmerged surfaces, but this isn't quite how it works in practice. 

### The implicit complement from the perspective of volume 1:

In this case the implicit complement is bounded by surface 1.

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/7ca5f588-9862-4e1d-bb12-a4909629fdf9">

### Implicit complement from the perspective of volume 2:

In this case the implicit complement is bounded by surface 2.

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/54b30567-87be-4a70-9545-a96a4edb5089">

## How this affects the plotting process:

In OpenMC we check each volume in a DAGMC universe for point containment to establish the position of a particle. To perform this check, we fire a ray in any given direction from the particle's location in an attempt to find an _exiting_ intersection with the triangles of the DAGMC volume.

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/0f2d9054-c7f1-49cd-ba3c-9fe1eb0075c2">

As we get into the subtleties of the problem, we'll recall that these surfaces are coincident and represent them as such in the image:

<img width="677" alt="image" src="https://github.com/openmc-dev/openmc/assets/4563941/4fb6c845-d9e5-47fe-96f4-44dd4adcba57">

While looping over volumes searching for a successful point containment check, it's possible that the implicit complement may be checked before volume 1 and 2. This is where the problem occurs. Because both surfaces 1 and 2 are part of the implicit complement and are coincident, a nearest intersection with either surface is valid and results in two outcomes:

  1. The particle ray intersects with surface 1: Based on the image of the IPC from the perspective of volume 1, this is considered to be an _entering_ intersection as if the particle were moving from volume 1 to the IPC volume and therefore the point is considered to be outside the IPC. The search loop continues and eventually identifies that the particle is in volume 1 (only the triangles of surface 2 are part of this search, so there's no ambiguity in that case)
  2. The particle ray intersects with surface 2: Based on the image of the IPC from the perspective of volume 2, this is considered to be an exiting intersection as if the particle were moving from the IPC to volume 2. The particle is considered to be in the IPC and the search loop stops **resulting in the artifacts we see in the plotter**.

### A proposed solution

Instead of including the IPC volume in the initial search for a containing cell, we can first check all of the intentionally created volumes in the CAD model and fall back to a check on the IPC at the end. In this way we still check all cells in the DAGMC universe, but we do not inadvertently exit the loop too soon and miss a particle placement. An implementation of this fix can be found [here](https://github.com/pshriwise/openmc/tree/dagmc-find-cell-updates) and does seem to remove these artifacts. Note there is some special behavior depending on whether or not the `DAGUniverse` is the root universe of the geometry or not.

![image](https://github.com/openmc-dev/openmc/assets/4563941/7871beaa-b12a-4c07-8b2d-780f7b3fa55d)

These changes requires the `DAGUniverse::find_cell` method to take ownership of the search loop (before we could rely on the `Universe::find_cell` routine). That was nice, but I don't see a problem with having a universe containing a fundamentally different geometry type have its own search loop.

---

I thought I'd tackle a couple of questions that could come to mind reading the post above.
### Why does this occur in some locations of a volume but not others? Why is it direction-dependent?

The intersection with coincident surfaces of triangles is going to come down to traversal order of the bounding volume hierarchy and floating point arithmetic during the triangle intersection calculation -- slight changes in the angle and position will result in selecting triangles from one surface or the other of the coincident surfaces. From the perspective of the ray tracer, a nearest intersection with either surface is valid.

### How might this adversely affect transport (after a particle is sourced)?

As long as the particle starts in the correct volume, the unmerged surfaces shouldn't have a negative effect during transport due to how intersections are filtered. During point containment checks, DAGMC searches for the nearest _exiting or entering_ intersection. During transport, DAGMC performs a query to find the nearest _exiting_ intersection. So in the example above as a particle moves out of volume 1 it will:

  1. Intersect with surface 1
  2. Cross into the IPC volume
  3. Intersect with surface 2 at a zero distance (intersections with surface 1 are filtered b/c they would be entering intersections)
  4. Cross into volume 2

So while there are additional steps involved, the unmerged surfaces shouldn't cause a problem with transport.

### Why didn't we see this in testing?

The test models were created as analogs of eigenvalue problems. As a result the geometries are simple and don't contain unmerged surfaces. Going forward, it's worth considering adding some models with unmerged surfaces to ensure we can handle these cases as expected -- though this is arguably something that should be part of DAGMC testing as opposed to a downstream application.

**Comment by gonuke** [comment]
Thanks for the great detective work here @pshriwise !!!  I'm sure that was hard to track down.  It's also comforting to know that it's a plotting issue and not a transport issue.

**Comment by paulromano** [comment]
Seriously, thanks for this essay @pshriwise, complete with a FAQ! I'm still a little confused about the transport question. If the plotter is telling us that `find_cell` results in certain locations being assigned to the implicit complement, wouldn't the same be true of source particles? That is, won't we have some source particles that are born in implicit complement when they shouldn't be?

**Comment by paulromano** [comment]
Also, given the reports in the [discourse group](https://openmc.discourse.group/t/weird-artifacts-on-dagmc-geometry-plots/3278), does this imply that the Cubit meshing is not imprinting/merging correctly?

**Comment by pshriwise** [comment]
> Seriously, thanks for this essay @pshriwise, complete with a FAQ! I'm still a little confused about the transport question. If the plotter is telling us that `find_cell` results in certain locations being assigned to the implicit complement, wouldn't the same be true of source particles? That is, won't we have some source particles that are born in implicit complement when they shouldn't be?

Yeah, that's entirely possible. That question was more meant to address the particle in flight after it is sourced. I updated the text to clarify.

**Comment by ebknudsen** [comment]
Thanks for the thorough explanation @pshriwise. Very very helpful. On the topic of transport: One thing that I have seen for geometries such as (pretty much) the one you describe, i.e. coincident surfaces where the "outside" is the IPC, is an increased leakage fraction. From what I gather this is consistent with the notion of IPC as you describe for the plotter.

**Comment by nplinden** [comment]
Hi all,
thanks for the work Patrick! It is very insightful.
I tried pulling your branch and re-plot my pump (not the teapot, since the teapot plot was faulty because the plot slice fell right on the surface boundary of two cells) , I get the following result :

![plot_2](https://github.com/openmc-dev/openmc/assets/57541749/bd850919-5783-4c2a-a843-ebe72b3772eb)

It seems that it didn't solve the issue for me unfortunately.
I did run a `openmc --version` to check that the git SHA1 corresponds the your last commit, which confirmed that I was indeed running your code and I was not mixing up my openmc binaries.
Do you think this could be a different issue altogether?

**Comment by pshriwise** [comment]
Sigh, more nuance here I think @Nicolas Linden. Did some plotting using DAGMC both with and without [double-down](https://github.com/pshriwise/double-down) enabled. double-down primarily enables Embree as a backend ray tracer for DAGMC, but it does perform the point containment check slightly differently than DAGMC. 

Without double-down enabled, I did see some effects like the ones above, but with double-down enabled I got clean images.

![xy](https://github.com/openmc-dev/openmc/assets/4563941/7cd6e82d-f179-4475-8223-33e803f1e1ac)

![yz](https://github.com/openmc-dev/openmc/assets/4563941/4f2db369-7362-44bd-93ef-902b38641066)

I also walked through it a bit in the plotter, but I didn't run into any artifacts there either.

![inplotter-xy](https://github.com/openmc-dev/openmc/assets/4563941/c0134995-f6a2-46e0-b3e7-5771fd3a12b4)

![inplotter-xz](https://github.com/openmc-dev/openmc/assets/4563941/09eceb85-889d-4840-923a-0910140d0732)

**Comment by nplinden** [comment]
Thanks for taking a look @pshriwise , this is very puzzling.
Do you think this could have some impact on transport ?

**Comment by pshriwise** [comment]
> Thanks for taking a look @pshriwise , this is very puzzling. Do you think this could have some impact on transport ?

It's likely, but I'm not entirely sure yet. I'll need to dig into it further.

**Comment by paulromano** [comment]
I believe with #2816 and #2838 merged, there should no longer be any plotting artifacts.
